README.md
# SQift

[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/SQift.svg)](https://img.shields.io/cocoapods/v/SQift.svg)
[![Carthage Compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![Platform](https://img.shields.io/cocoapods/p/SQift.svg?style=flat)](http://cocoadocs.org/docsets/SQift)

SQift is a lightweight Swift wrapper for SQLite.

## Features

- [X] On-Disk, In-Memory and Temporary Database Connections
- [X] SQL Statement Execution
- [X] Generic Parameter Binding and Value Extraction
- [X] Codable and Codable Collection Bindings
- [X] Simple Query APIs for Values, Rows, and Collections
- [X] Transactions and Savepoints
- [X] Tracing and Trace Event Support
- [X] Scalar and Aggregate Functions
- [X] Commit, Rollback, Update, Authorizer Hooks
- [X] WAL Checkpointing
- [X] ConnectionQueue for Serial Execution per Database Connection
- [X] ConnectionPool for Parallel Execution of Read-Only Connections
- [X] Top-Level Database to Simplify Thread-Safe Reads and Writes
- [X] Database Migration Support
- [X] Database Backups
- [x] Comprehensive Unit Test Coverage
- [x] Complete Documentation

## WildThink Extensions

- [X] Shared Memory Option
- [ ] JSON Support
- [ ] AppDatabase
- [ ] Database Snapshots

## Requirements

- iOS 10.0+, macOS 10.12+, tvOS 10.0+, watchOS 3.0+
- Xcode 10.2+
- Swift 5.0+

## Migration Guides

- [SQift 4.0 Migration Guide](Documentation/SQift%204.0%20Migration%20Guide.md)

## Communication

- Need help? Open an issue.
- Have a feature request? Open an issue.
- Find a bug? Open an issue.
- Want to contribute? Fork the repo and submit a pull request.

---

## Installation

### Swift Package Manager

The [Swift Package Manager](https://swift.org/package-manager/) is a tool for managing the distribution of Swift code. It’s integrated with the Swift build system to automate the process of downloading, compiling, and linking dependencies.

>Xcode 11+ is required to provide SwiftPM support for iOS, watchOS, and tvOS platforms.

In Xcode menu `File -> Swift Packages -> Add Package Dependency...` enter repository URL `https://github.com/wildthink/SQift.git`.

### CocoaPods

[CocoaPods](http://cocoapods.org) is a dependency manager for Cocoa projects.
You can install it with the following command:

```bash
$ gem install cocoapods
```

> CocoaPods 1.3+ is required to build SQift.

To integrate SQift into your Xcode project using CocoaPods, specify it in your `Podfile`:

```ruby
platform :ios, '11.0'
use_frameworks!

target '<Your Target Name>' do
    pod 'SQift', '~> 4.0'
end
```

Then, run the following command:

```bash
$ pod install
```

### Carthage

[Carthage](https://github.com/Carthage/Carthage) is a decentralized dependency manager that builds your dependencies and provides you with binary frameworks.

You can install Carthage with [Homebrew](http://brew.sh/) using the following command:

```bash
$ brew update
$ brew install carthage
```

To integrate SQift into your Xcode project using Carthage, specify it in your [Cartfile](https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#cartfile):

```
github "wildthink/SQift" ~> 5.0
```

Run `carthage update` to build the framework and drag the built `SQift.framework` into your Xcode project.

---

## Usage

SQift is designed to make it as easy as possible to work with SQLite from Swift.
It does not, however, eliminate the need to understand how SQLite actually works.
Before diving into SQift, it is recommended to first have a firm grasp on what SQLite is, how it works and how to use it.

- [SQLite Docs](https://www.sqlite.org/docs.html)
- [SQLite Tutorial](http://zetcode.com/db/sqlite/)

SQift heavily leverages the new error handling model released with Swift 2.0.
It was designed from the ground up to throw in all applicable cases.
This makes it easy to wrap all your SQift calls in the do-catch paradigm.

### Creating a Database Connection

Creating a database connection is simple.

```swift
let onDiskConnection = try Connection(storageLocation: .onDisk("path_to_db"))
let inMemoryConnection = try Connection(storageLocation: .inMemory)
let tempConnection = try Connection(storageLocation: .temporary)
```

There are also convenience parameters to make it easy to customize the flags when initializing the database connection:

```swift
let connection = try Connection(
    storageLocation: .onDisk("path_to_db"),
    readOnly: true,
    multiThreaded: false,
    sharedCache: false
)
```

> In most cases, the default values are preferred.
> For more details about creating a database connection, please refer to the SQLite [documentation](https://www.sqlite.org/c3ref/open.html).

### Executing Statements

To execute a SQL statement on the `Connection`, you need to first create a `Connection`, then call `execute`.

```swift
let connection = try Connection(storageLocation: .onDisk("path_to_db"))

try connection.execute("PRAGMA foreign_keys = true")
try connection.execute("PRAGMA journal_mode = WAL")

try connection.execute("CREATE TABLE cars(id INTEGER PRIMARY KEY, name TEXT, price INTEGER)")

try connection.execute("INSERT INTO cars VALUES(1, 'Audi', 52642)")
try connection.execute("INSERT INTO cars VALUES(2, 'Mercedes', 57127)")

try connection.execute("UPDATE cars SET name = 'Honda' where id = 1")
try connection.execute("UPDATE cars SET price = 61_999 where name = 'Mercedes'")

try connection.execute("DELETE FROM cars where name = 'Mercedes'")
try connection.execute("DROP TABLE cars")
```

### Bindings

Most Swift data types cannot be directly stored inside the database.
They need to be converted to a data type supported by SQLite.
In order to support moving Swift data types into the database and back out again, SQift leverage three powerful protocols: `Bindable`, `Extractable`, and `Binding`.

#### Bindable Protocol

The `Bindable` protocol handles converting Swift data types into a `BindingValue` enumeration type which can be stored in the database.

```swift
public protocol Bindable {
    var bindingValue: BindingValue { get }
}
```

#### Extractable Protocol

While the `Bindable` protocol helps move Swift data types into the database, the `Extractable` protocol allows SQift to extract the values from the database `Connection` and convert them back to the requested Swift data type.

```swift
public protocol Extractable {
    typealias BindingType
    typealias DataType = Self
    static func fromBindingValue(_ value: Any) -> DataType?
}
```

#### Binding Protocol

To extend Swift data types to be able to be inserted into the database and also be extracted safely, the `Binding` protocol forces the data type to conform to both the `Bindable` and `Extractable` protocols.

```swift
public protocol Binding: Bindable, Extractable {}
```

In order to make it as easy as possible to use SQift, SQift extends the following Swift data types to conform to the `Binding` protocol:

- **NULL:** `NSNull`
- **INTEGER:** `Bool`, `Int8`, `Int16`, `Int32`, `Int64`, `Int`, `UInt8`, `UInt16`, `UInt32`, `UInt64`, `UInt`
- **REAL:** `Float`, `Double`
- **TEXT:** `String`, `URL`, `Date`
- **BLOB:** `Data`

> Additional Swift data types can easily add `Bindable` protocol conformance if necessary.

### Binding Parameters to a Statement

Safely binding parameters to a `Statement` is easy thanks to the `Binding` protocol.
First you need to `prepare` a `Statement` object, then `bind` the parameters and `run` it using method chaining.

```swift
let connection = try Connection(storageLocation: .onDisk("path_to_db"))

try connection.prepare("INSERT INTO cars VALUES(?, ?, ?)").bind(1, "Audi", 52_642).run()
try connection.prepare("INSERT INTO cars VALUES(:id, :name, :price)").bind([":id": 1, ":name": "Audi", ":price": 52_642]).run()
```

There are also convenience methods on the `Connection` for preparing a `Statement`, binding parameters and running it all in a single method named `run`.

```swift
let connection = try Connection(storageLocation: .onDisk("path_to_db"))

try connection.run("INSERT INTO cars VALUES(?, ?, ?)", 1, "Audi", 52_642)
try connection.run("INSERT INTO cars VALUES(:id, :name, :price)", parameters: [":id": 1, ":name": "Audi", ":price": 52_642])
```

> It is very important to properly escape all parameter values provided in a SQL statement.
> When in doubt, always use the provided bind functionality.

### Querying Data

Querying data from the database makes extensive use of the `Binding` protocol.
It extracts the original values from the database, then uses the `Binding` protocol along with generics to convert the final Swift type.

#### Single Values

Extracting a single value from the database can be done using the `query` API.

```swift
let synchronous: Int? = try db.query("PRAGMA synchronous")
let minPrice: UInt? = try db.query("SELECT avg(price) FROM cars WHERE price > ?", 40_000)
```

#### Multiple Values

You can also use the `query` API to extract multiple values through the `Row` type.

```swift
if let row = try db.query("SELECT name, type, price FROM cars WHERE type = ? LIMIT 1", "Sedan") {
    let name: String = row[0]
    let type: String = row[1]
    let price: UInt = row[2]
}
```

The values can be accessed by index or by name.

```swift
if let row = try db.query("SELECT name, type, price FROM cars WHERE type = ? LIMIT 1", "Sedan") {
    let name: String = row["name"]
    let type: String = row["type"]
    let price: UInt = row["price"]
}
```

> The `Row` type supports both optional and non-optional value extraction through index and name subscripts.
> The non-optional subscripts are certainly the safest, but not always the most convenient.
> It is up to you to decide which makes more sense to use in each situation.
> Generally, try to use the optional types where the SQL statement is decoupled from the row value extraction.
> In these cases, the `ExpressibleByRow` type can be useful to help handle optionality.

#### ExpressibleByRow Types

In many cases, you want to construct model objects from a row in a result set.
The `ExpressibleByRow` type was designed for this use case.

```swift
protocol ExpressibleByRow {
    init(row: Row) throws
}
```

To make use of the `ExpressibleByRow` protocol, first create your model object and conform to the protocol.

```swift
struct Car {
    let name: String
    let type: String
    let price: UInt
}

extension Car: ExpressibleByRow {
    init(row: Row) throws {
        guard
            let name: String = row[0],
            let type: String = row[1],
            let price: UInt = row[2]
        else {
            throw ExpressibleByRowError(type: Car.self, row: row)
        }

        self.name = name
        self.type = type
        self.price = price
    }
}
```

Then, you can use the `query` API to automatically convert the result set into a `Car`.

```swift
let car: Car? = try db.query("SELECT name, type, price FROM cars WHERE type = ? LIMIT 1", "Sedan")
```

#### Multiple Rows

The `query` APIs also support querying for result sets with multiple rows.

```swift
let names: [String] = try db.query("SELECT name FROM cars")
let cars: [Car] = try db.query("SELECT * FROM cars WHERE price > ?", [20_000])
```

In addition to array result sets, you can also create dictionary result sets.

```swift
let sql = "SELECT name, price FROM cars WHERE price > ?"
let prices: [String: UInt] = try db.query(sql, 20_000) { ($0[0], $0[1]) }
```

### Dates

SQift has full binding support for `Date` objects which allows you to easily leverage the builtin date functionality in SQLite.
You can insert dates easily into the database, run queries against them, and then extract them back out.
SQift handles this through the `Date` binding which leverages the `bindingDateFormatter`.
By default, all `Date` types are stored in the database as `TEXT`, so make sure you map your column types accordingly.

```swift
let date1975: Date!
let date1983: Date!
let date1992: Date!
let date2001: Date!

try connection.execute(
	"CREATE TABLE cars(id INTEGER PRIMARY KEY, name TEXT, release_date TEXT)"
)

try connection.execute("INSERT INTO cars(?, ?)", "70s car", date1975)
try connection.execute("INSERT INTO cars(?, ?)", "80s car", date1983)
try connection.execute("INSERT INTO cars(?, ?)", "90s car", date1992)
try connection.execute("INSERT INTO cars(?, ?)", "00s car", date2001)
```

Once you have your dates stored in the database, you can run date range queries to narrow down your data.

```swift
let date1980: Date!
let date2000: Date!

let carCount: Int = try connection.query(
    "SELECT count(*) FROM cars WHERE release_date >= date(?) AND release_date <= date(?)",
    date1980, 
    date2000
)

print("Total Cars from the 80s and 90s: \(carCount)") // should equal 2
```

> You can swap the default date formatting, but be careful when doing so.
> You need to make sure the new date format complies with the SQLite [requirements](https://www.sqlite.org/lang_datefunc.html) so date range queries will continue to work as expected.

You can also extract dates out of each row as an `Date`.

```swift
let sql = "SELECT release_date WHERE name = ? LIMIT 1"
let releaseDate: Date = try connection.query(sql, "80s car")
```

### Custom Bindings

SQift has support for many common primitive types, but what about when you want to store a custom type in the database?
This is where custom bindings come into play.
All you need to do to store your own custom types in the database is conform to the `Binding` protocol.

```swift
enum DownloadState: Int {
    case pending, downloading, downloaded, failed
}

extension DownloadState: Binding {
    typealias BindingType = Int64

    var bindingValue: BindingValue { return .integer(Int64(rawValue)) }

    static func fromBindingValue(_ value: Any) -> AssetType? {
        guard let value = value as? Int64, let rawValue = Int(exactly: value) else { return nil }
        return DownloadState(rawValue: rawValue)
    }
}

try connection.execute("CREATE TABLE downloads(name TEXT PRIMARY KEY, state INTEGER NOT NULL)")
try connection.run("INSERT INTO downloads VALUES(?, ?)", "image1", DownloadState.pending)

if let state: DownloadState? = try connection.query("SELECT state FROM downloads WHERE name = 'image1') {
    print(state)
}

// Output
// DownloadState.pending
```

#### Codable Bindings

SQift also supports `Codable` bindings out-of-the-box.
For example, let's say we have a `Person` object that is `Codable`.
We can store `Person` instances directly in the database without having to create a custom binding.

```swift
struct Employee {
    let id: Int64
    let firstName: String
    let lastName: String
    let age: UInt
}

let phil = Person(id: 1, firstName: "Phil", lastName: "Knight", age: 79)

try connection.execute("CREATE TABLE employees(id INTEGER PRIMARY KEY, employee BLOB NOT NULL)")
try connection.run("INSERT INTO employees(employee) VALUES(?)", phil)

if let employee1: Employee? = try connection.query("SELECT employee FROM employees WHERE id = 1) {
    print(employee.firstName)
}

// Output
// "phil"
```

> You need to think through whether it makes sense for you to use `Codable` bindings or not.
> While they are very convenient, the information inside them cannot be queried.
> In the above example, you could not run a query such as `SELECT count(1) FROM employees where firstName = 'Phil'`.
> If your use case does not require you to run such a search, then `Codable` bindings may be a useful choice.

#### Codable Collections

SQift also supports `Codable` collections through the `ArrayBinding`, `SetBinding`, and `DictionaryBinding` types.

```swift
let points: ArrayBinding = [
    CGPoint(x: 1.0, y: 2.0),
    CGPoint(x: 3.0, y: 4.0),
    CGPoint(x: 5.0, y: 6.0),
    CGPoint(x: 7.0, y: 8.0),
    CGPoint(x: 9.0, y: 10.0)
]

try connection.execute("CREATE TABLE stream(id INTEGER PRIMARY KEY, data BLOB NOT NULL)")
try connection.run("INSERT INTO stream(data) VALUES(?)", points)

let pointsQueried: ArrayBinding<CGPoint>? = try connection.query("SELECT data FROM stream WHERE id = 1")

pointsQueried?.elements.forEach { print("(\($0.x), \($0.y))") }

// Output
// (1.0, 2.0)
// (3.0, 4.0)
// (5.0, 6.0)
// (7.0, 8.0)
// (9.0, 10.0)
```

> Codable collections can be useful in situations where you are writing large streams of data that are never partially queried.
> If you only write the data in one pass, and only query the data as the entire set, codable collections might be a good option.

### Transactions

Changes cannot be made to the database except within a transaction.
By default, any command that changes the database will automatically start a transaction if one is not already in effect.
Transactions can also be started manually in SQift when multiple operations need to be run inside a single transaction.

```swift
try connection.execute("CREATE TABLE cars(id INTEGER PRIMARY KEY, name TEXT, price INTEGER)")

try connection.transaction {
    try connection.prepare("INSERT INTO cars VALUES(?, ?, ?)").bind(1, "Audi", 52642).run()
    try connection.prepare("INSERT IN cars VALUES(?, ?, ?)").bind(2, "Mercedes", 57127).run()
}
```

> If any error occurs within the transaction, all the changes are automatically rolled back by SQift.

### Tracing

When debugging SQL statements, sometimes it can be helpful to be able to print out what is actually being executed by SQLite.
SQift allows you to do this through the `traceEvent` API by registering a closure to run for each statement execution.

```swift
let connection = try Connection(storageLocation: storageLocation)

connection.traceEvent { event in
    if case .statement(_, let sql) = sql {
        print(sql)
    }
}

try connection.execute("CREATE TABLE employees(id INTEGER PRIMARY KEY, name TEXT)")
try connection.prepare("INSERT INTO employees VALUES(?, ?)").bind(1, "Bill Bowerman").run()
try connection.prepare("INSERT INTO employees VALUES(?, ?)").bind(2, "Phil Knight").run()
let employees: [Employee] = try connection.query("SELECT * FROM employees")

// Output
// "CREATE TABLE employees(id INTEGER PRIMARY KEY, name TEXT)"
// "INSERT INTO employees VALUES(1, 'Bill Bowerman')"
// "INSERT INTO employees VALUES(2, 'Phil Knight')"
// "SELECT * FROM employees"
```

> The `traceEvent` API allows you to be more selective about the types of statements you want to trace.
> You can select which types of statements you want by using the trace event masks.

### Collations

SQift supports custom collation functions for cases where the three built-in collating functions are not sufficient.
A couple real-world examples of custom cases might include: diacritic aware sorting and numerical sorting.

```swift
let connection = try Connection(storageLocation: storageLocation)

connection.createCollation(named: "NUMERIC") { lhs, rhs in
    return lhs.compare(rhs, options: .numeric, locale: .autoupdatingCurrent)
}

try connection.execute("CREATE TABLE values(text TEXT COLLATE 'NUMERIC' NOT NULL)")

let values = ["string 1", "string 21", "string 12", "string 11", "string 02"]

try values.forEach { try connection.run("INSERT INTO values(text) VALUES(?)", $0) }
let extractedValues: [String] = try connection.query("SELECT * FROM values ORDER BY text")

extractedValues.forEach { print($0) }

// Output
// "string 1"
// "string 02"
// "string 11"
// "string 12"
// "string 21"
```

### Functions

While SQLite is a very robust library, sometimes you will run into cases where you need to extend the functionality of SQLite where it is limited.
For example, you may need to create a custom function to determine what month of a calendar year a particular date falls within.
SQLite cannot do this directly since it lacks calendar support.

SQift supports custom scalar and aggregate functions.
The following is a simple example of how you could extend SQLite to support a `strip_unicode` function.

```swift
try connection.addScalarFunction(named: "strip_unicode", argumentCount: 1) { _, values in
    guard
        let value = values.first, value.isText,
        let valueData = value.text.data(using: .ascii, allowLossyConversion: true),
        let asciiValue = String(data: valueData, encoding: .ascii)
    else { return .null }

    return .text(asciiValue)
}

let sql = "SELECT strip_unicode(?)"

let result1: String? = try connection.prepare(sql, "å").query()
let result2: String? = try connection.prepare(sql, "ć").query()
let result3: String? = try connection.prepare(sql, "áč").query()

// result1 = "a"
// result2 = "c"
// result3 = "ac"
```

> For more advanced examples of scalar and aggregate functions, please refer to the test suite.

---

## Advanced

### Hooks

SQift has support built in for commit, rollback, update, and authorizer hooks.
The commit hook is used to determine whether a commit should be executed or rolled back.
The rollback hook is called when a commit is rolled back.

```swift
var shouldCancelCommit = false

connection.commitHook { return shouldCancelCommit }
connection.rollbackHook { print("rollback occurred") }
```

Update hooks can be used to react to `.insert`, `.update`, or `.delete` operations.

```swift
connection.updateHook { type, databaseName, tableName, rowID in
    var message = "\(type) row \(rowID)"

    if let databaseName = databaseName { message += " on \(databaseName)" }
    if let tableName = tableName { message += ".\(tableName)" }

    print(message)
    
    // Could update the file system, invalidate a cache, send notifications, etc.
}

try connection.execute("""
    INSERT INTO employee(name) VALUES('Phil Knight');
    UPDATE person SET name = 'Bill Bowerman' WHERE id = 1;
    DELETE FROM person WHERE id = 1
    """
)

// Output
// "insert row 1 on main.person"
// "update row 1 on main.person"
// "delete row 1 on main.person"
```

The authorizer hook is the most complex of the four.
It allows you to control what statements are allowed to run on a connection.
For example, you could disable all actions on a particular connection other than select statements.

```swift
try connection.authorizer { action, p1, p2, p3, p4 in
    guard action == .select else { return .deny }
    return .ok
}
```

### Checkpoints

Databases with a `WAL` journal mode use checkpoint operations to move updates from the WAL file into the database.
SQift supports checkpoints and busy timeouts and handlers which can be useful in certain situations.
For example, you may want to use a `WAL` database for performance reasons, then transfer it to a remote server or different device.
Before doing this, it is wise to checkpoint the database and also vacuum it.

```swift
try connection.busyHandler(.timeout(1.0)) // 1 second
let checkpointResult = try connection.checkpoint(mode: .truncate)

try connection.execute("VACUUM")
```

> Checkpointing is a very complex process.
> Before using the `checkpoint` APIs, make sure to read through the SQLite [documentation](https://sqlite.org/c3ref/wal_checkpoint_v2.html).

### Thread Safety

Thread-safety is a complex topic when it comes to SQLite.
As a general rule, it is NEVER safe to access a database `Connection` from multiple threads simultaneously.
Each connection should be accessed serially to guarantee safety. 

If you wish to access the database in parallel, there are a few things you need to know.
First off, you'll need to use Write-Ahead Logging by setting the journal mode to `WAL`.
By changing the database to a `WAL` journal mode, the database can be read during a write and written during a read in parallel using multiple connections.

```swift
try connection.execute("PRAGMA journal_mode = WAL")
```

Another important note is that SQLite can only perform write operations serially, no matter how many connections you create.
Therefore, you should only ever create a single connection for writing if possible.
You can use as many reader connections as you wish.
For more information about thread-safety and WAL journal modes, please refer to the following:

- [Write-Ahead Logging](https://www.sqlite.org/wal.html)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
`SQift` adheres to [Semantic Versioning](http://semver.org/).

#### 5.x Releases

- `5.1.x` Releases - [5.1.0](#510) 
- `5.0.x` Releases - [5.0.0](#500) | [5.0.1](#501)

#### 4.x Releases

- `4.1.x` Releases - [4.1.0](#410)
- `4.0.x` Releases - [4.0.0](#400)

#### 3.x Releases

- `3.3.x` Releases - [3.3.0](#330)
- `3.2.x` Releases - [3.2.0](#320) | | [3.2.1](#321)
- `3.1.x` Releases - [3.1.0](#310)
- `3.0.x` Releases - [3.0.0](#300) | [3.0.1](#301)

#### 2.x Releases

- `2.1.x` Releases - [2.1.0](#210) | [2.1.1](#211)
- `2.0.x` Releases - [2.0.0](#200)

#### 1.x Releases

- `1.1.x` Releases - [1.1.0](#110) | [1.1.1](#111) | [1.1.2](#112)
- `1.0.x` Releases - [1.0.0](#100)

#### 0.x Releases

- `0.9.x` Releases - [0.9.0](#090)
- `0.8.x` Releases - [0.8.0](#080)
- `0.7.x` Releases - [0.7.0](#070) | [0.7.1](#071)
- `0.6.x` Releases - [0.6.0](#060) | [0.6.1](#061)
- `0.5.x` Releases - [0.5.0](#050)
- `0.4.x` Releases - [0.4.0](#040)
- `0.3.x` Releases - [0.3.0](#030)
- `0.2.x` Releases - [0.2.0](#020)
- `0.1.x` Releases - [0.1.0](#010)

---
## [5.1.0](https://github.com/wildthink/SQift/releases/tag/5.1.0)
#### WildThink Beta 1
- Added `StorageLocation.sharedMemory()` option so all Database Connections share the same database.
- Changed default `Database StorageLocation` to `.sharedMemory` and disallow `.inMemory` option.
- Added basic SQLite JSON Support (see [JSON1 Extension](https://www.sqlite.org/json1.html))
- Changed `Database` to `open` to subclassing.
- Added `AppDatabase`.

## [5.0.1](https://github.com/Nike-Inc/SQift/releases/tag/5.0.1)

Released on 2019-10-25. All issues associated with this milestone can be found using this
[filter](https://github.com/Nike-Inc/SQift/milestone/12?closed=1).

#### Fixed

- Issue where Carthage was unable to build SQift due to schemes requiring test suite compilation.
  - Fixed by [William Thompson](https://github.com/eSpecialized) in Pull Request [#28](https://github.com/Nike-Inc/SQift/pull/28).
- Issue in Xcode project Documentation group where links were invalid and README links were broken.
  - Fixed by [James Eisenhauer](https://github.com/jimmyeisenhauer) in Pull Request [#30](https://github.com/Nike-Inc/SQift/pull/30).
- Compiler warnings in `Function.swift` picked up by Xcode 11.
   - Fixed by [Christian Noon](https://github.com/cnoon) in Pull Request [#31](https://github.com/Nike-Inc/SQift/pull/31).

## [5.0.0](https://github.com/Nike-Inc/SQift/releases/tag/5.0.0)

Released on 2019-04-30. All issues associated with this milestone can be found using this
[filter](https://github.com/Nike-Inc/SQift/milestone/10?closed=1).

#### Updated

-  Dropped Swift 4.2 support.
   - Updated by [William Thompson](https://github.com/eSpecialized) in Pull Request [#19](https://github.com/Nike-Inc/SQift/pull/19).

---

## [4.1.0](https://github.com/Nike-Inc/SQift/releases/tag/4.1.0)

Released on 2019-04-30. All issues associated with this milestone can be found using this
[filter](https://github.com/Nike-Inc/SQift/milestone/8?closed=1).

#### Updated

- To Swift 5 with backwards compatibility to Swift 4.2.
   - Updated by [William Thompson](https://github.com/eSpecialized) in Pull Request [#18](https://github.com/Nike-Inc/SQift/pull/18).
- Deployment targets to watchOS 3.0 and tvOS 10.0 in Xcode project to match podspec.
   - Updated by [William Thompson](https://github.com/eSpecialized) in Pull Request [#18](https://github.com/Nike-Inc/SQift/pull/18).

---

## [4.0.0](https://github.com/Nike-Inc/SQift/releases/tag/4.0.0)

Released on 2018-11-30. All issues associated with this milestone can be found using this
[filter](https://github.com/Nike-Inc/SQift/milestone/7?closed=1).

#### Added

- Migration Guide for SQift 4 and added it to the README.
  - Added by [Jereme Claussen](https://github.com/jereme) in Pull Request [#16](https://github.com/Nike-Inc/SQift/pull/16).

#### Updated

- Deployment targets to iOS 9.0, Mac OSX 10.11, WatchOS 2.0 and tvOS 9.0.
  - Updated by [Jereme Claussen](https://github.com/jereme) in Pull Request [#16](https://github.com/Nike-Inc/SQift/pull/16).

#### Removed

- `Connection.trace` in favor of `Connection.traceEvent()`
  - Removed by [Jereme Claussen](https://github.com/jereme) in Pull Request [#16](https://github.com/Nike-Inc/SQift/pull/16).

---

## [3.3.0](https://github.com/Nike-Inc/SQift/releases/tag/3.3.0)

Released on 2018-11-30. All issues associated with this milestone can be found using this
[filter](https://github.com/Nike-Inc/SQift/milestone/5?closed=1).

#### Deprecated

- The `trace` API to prepare for 4.0 migration since the internal `sqlite3_trace` API is deprecated when bumping deployment targets.
  - Deprecated by [Christian Noon](https://github.com/cnoon) in Pull Request
  [#15](https://github.com/Nike-Inc/SQift/pull/15).

---

## [3.2.1](https://github.com/Nike-Inc/SQift/releases/tag/3.2.1)

Released on 2018-09-17. All issues associated with this milestone can be found using this
[filter](https://github.com/Nike-Inc/SQift/milestone/6?closed=1).

#### Added

- An internal initializer to `SQLiteError` to assist in error mocking for testing.
  - Added by [Christian Noon](https://github.com/cnoon).

#### Updated

- The Xcode project to use Xcode 9.3 compatibility mode.
  - Updated by [Christian Noon](https://github.com/cnoon) in Pull Request
  [#10](https://github.com/Nike-Inc/SQift/pull/10).
- The Xcode project to support Xcode 10 and Swift 4.2.
  - Updated by [Christian Noon](https://github.com/cnoon) in Pull Request
  [#10](https://github.com/Nike-Inc/SQift/pull/10).
- The Travis-CI YAML file to use `bundler` and Xcode 10.
  - Updated by [Christian Noon](https://github.com/cnoon) in Pull Request
  [#10](https://github.com/Nike-Inc/SQift/pull/10).

## [3.2.0](https://github.com/Nike-Inc/SQift/releases/tag/3.2.0)

Released on 2018-05-02. All issues associated with this milestone can be found using this

SQiftLab.playground/Contents.swift
import Cocoa
import SQift

// https://news.ycombinator.com/item?id=19277809
// https://github.com/rgov/sqlite_protobuf
// https://www.sqlite.org/json1.html#jgrouparray
// https://gist.github.com/wildthink/287848614e9bc4c984357d3c72d7479d


struct Location: CodableBinding {
    typealias BindingType = String
    var name: String
}

let dbc = try Connection(storageLocation: .inMemory)

try dbc.execute("CREATE TABLE cars(id INTEGER PRIMARY KEY, name TEXT, price INTEGER, tags JSON, location JSON, freq TEXT)")

let tags = """
["alpha", "beta", 33]
"""

let itags = """
[1,2,3]
"""

let loc = Location(name: "Somewhere")
var loc_s = """
{"name":"Somewhere"}
"""

if case BindingValue.text(let bv_s) = loc.bindingValue {
    print (#line, bv_s)
    loc_s = bv_s
}

let statement = try dbc.prepare("INSERT INTO cars(name, price, tags, location) VALUES(?, ?, ?, ?)", "Ford Focus", 25_999, tags, loc_s)

try statement.bind("Ford Sprint", 28_999, tags, loc_s).run()

try dbc.execute("INSERT INTO cars (name, price, tags) VALUES('Audi', 52642, '\(tags)')")
try dbc.execute("INSERT INTO cars (name, price, tags) VALUES('Mercedes', 57127, '\(tags)')")
try dbc.execute("INSERT INTO cars (name, price, tags) VALUES('Ford', 20000, '\(tags)')")
try dbc.execute("INSERT INTO cars (name, price, tags, freq) VALUES ('Mazda', 25000, '\(tags)', 'anytime')")

// The struct need NOT be Codable but must be JSON compatible
public struct Car {
    public enum Frequency: String, Codable, CaseIterable { case once, daily, weekly, anytime }

    let id: Int64
    let name: String
//    let type: String
    let price: UInt
    let tags: [Any] // [AnyCodabe]
    let location: Location?
    let frequency: Frequency?
    
//    let location: [String:AnyCodable]?
}

extension Car.Frequency: Extractable {
    public typealias BindingType = String
    public static func fromBindingValue(_ value: Any) -> Car.Frequency? {
        guard let key = value as? String else { return nil }
        for f in Car.Frequency.allCases {
            if f.rawValue == key { return f }
        }
        return nil
    }
}

extension Car: ExpressibleByRow {
    public init(row: Row) throws {
        self.id = row[0]
        self.name = row[1]
        self.price = row[2]
        self.tags = row[3]
        self.location = row[4]
        self.frequency = row[5]
    }
//    public init(row: Row) throws {
//        guard
//            let id: Int64 = row[0],
//            let name: String = row[1],
//            let price: UInt = row[2],
//            let tags: [AnyCodable] = row[3]
//            else {
//                throw ExpressibleByRowError(type: Car.self, row: row)
//        }
//        self.id = id
//        self.name = name
////        self.type = type
//        self.price = price
//        self.tags = tags
//        self.location = row[4]
//        self.frequency = row[5]
//    }
}

let jq_1: SQL = """
SELECT name, tags.value
 FROM cars, json_each(cars.tags) as tags
WHERE tags.value = 33
"""

let jq_2: SQL = """
select name, json_extract(cars.location, '$.name') from cars
"""

let jq_3: SQL = """
SELECT name
 FROM cars, json_each(cars.location, '$.name') as loc_name
WHERE loc_name.value LIKE 'Bum%'
"""

let jq_4: SQL = """
SELECT name, json_extract(tags, '$[0]') as jt
 FROM cars
"""

try dbc.query(jq_1)
let bname = try dbc.query(jq_2)
try dbc.fetch(jq_2, []) {
    print (#line, $0)
}

try dbc.query(jq_3)
try dbc.query(jq_4)
//print(#line, bname)

let names: [String] = try dbc.query("SELECT name FROM cars WHERE price >= ?", [20_000])

print (names)

// let cars: [Car] = try dbc.query("SELECT * FROM cars, json_each(cars.tags) as tag WHERE tag.value == 'alpha' OR price > ?", [20_000])
let cars: [Car] = try dbc.query("SELECT * FROM cars, json_each(cars.tags) as tags WHERE tags.value == 'alpha'", [])
print (#line, "CAR:", cars)

let stmnt = try dbc.prepare("SELECT * FROM cars WHERE price > ?", [20_000])

try stmnt.fetch {
    print ($0)
}

var rows: [Any] = []

try dbc.fetch("SELECT * FROM cars WHERE price >= ?", [20_000]) {
    print ($0)
    rows.append($0.values)
}

for row in rows {
    print (#line, row)
}

try dbc.fetch("PRAGMA table_info('cars')", []) {
    print ($0)
}

try dbc.fetch("select * from pragma_table_info('cars')", []) {
    print ($0)
}

//let db = try Database(storageLocation: .inMemory)
//let db = try Database(storageLocation: .onDisk("/Users/jason/demo.db"))
let db = try Database(storageLocation: .sharedMemory("demo"))
//let db = try Database(storageLocation: .temporary)

try db.executeWrite {
    try $0.execute("CREATE TABLE cars(id INTEGER PRIMARY KEY, name TEXT, price INTEGER, tags JSON, location JSON, freq TEXT)")

    try $0.execute("INSERT INTO cars (name, price, tags) VALUES('Audi', 52642, '\(tags)')")
    try $0.execute("INSERT INTO cars (name, price, tags) VALUES('Mercedes', 57127, '\(tags)')")
    print(#line, "PASS")
}

try db.executeRead {
    let row = try $0.query("SELECT * FROM cars WHERE price > ?", [20_000])
    let stmnt = try $0.prepare("SELECT * FROM cars", [])
    let count: Int? = try $0.query("SELECT count(*) FROM cars", [])

    try stmnt.fetch {
        print (#line, $0)
    }
}
print(#line, "PASS")

/*
func bindingValue(value: Any) -> String {
    
    var str: String

    do {
        let data = try JSONSerialization.data(withJSONObject: value, options: [])
        str = String(data: data, encoding: .utf8) ?? ""
    } catch {
        str = ""
    }

    return str
}

func fromBinding(_ value: Any) -> Any? {
    guard let value = value as? String else { return nil }
    guard let data = value.data(using: .utf8) else { return nil }
    let json = try? JSONSerialization.jsonObject(with: data, options: [])
    return json
}

let list = ["one", "two"] // [1, 2, 3, 4]
let bv = bindingValue(value: list)
let v = fromBinding(bv)

let plist: [String : AnyCodable] = ["a": 23, "b": "stirng" ]
let pv = plist.bindingValue

if case BindingValue.text(let bv_s) = pv {
    [String:AnyCodable].fromBindingValue(bv_s)
}
*/
